test/tools/cli/IonProcess.java

```diff
@@ -0,0 +1,314 @@
+                ionReader.stepIn();
+
+                //recursive call
+                processEvent(ionWriterForOutput, ionWriterForErrorReport, ionReader, curTable);
+
+                //write a Container_End event and step out
+                new Event(EventType.CONTAINER_END,curType,null,null,
+                        null,null,null, curDepth )
+                        .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+                ionReader.stepOut();
+            } else {
+                ionStreamToEvent(EventType.SCALAR,ionReader).writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+            }
+        }
+    }
+
+    /**
+     * this function is only used for CONTAINER_START and SCALAR eventType since other types don't need initial data.
+     */
+    private static Event ionStreamToEvent(EventType eventType, IonReader ionReader) throws Exception {
+        if(ionReader.getType() == null) return null;
+
+        IonType ionType = ionReader.getType();
+        SymbolToken field_name = ionReader.getFieldNameSymbol();
+        SymbolToken[] annotations = ionReader.getTypeAnnotationSymbols();
+
+        String value_text = null;
+        byte[] value_binary = null;
+
+        if(eventType == EventType.SCALAR) {
+            //write Text
+            ByteArrayOutputStream textOut = new ByteArrayOutputStream();
+            IonWriter textWriter = IonTextWriterBuilder.standard().build(textOut);
+            textWriter.writeValue(ionReader);
+            textWriter.finish();
+            value_text = textOut.toString();
+            //write binary
+            ByteArrayOutputStream binaryOut = new ByteArrayOutputStream();
+            IonWriter binaryWriter = IonBinaryWriterBuilder.standard().build(binaryOut);
+            binaryWriter.writeValue(ionReader);
+            binaryWriter.finish();
+            value_binary = binaryOut.toByteArray();
+        }
+
+        ImportDescriptor[] imports = null;
+
+        int depth = ionReader.getDepth();
+
+        return new Event(eventType,ionType,field_name,annotations,value_text,value_binary,imports,depth);
+    }
+
+    private static boolean isSameSymbolTable(SymbolTable x, SymbolTable y){
+        if(x.isSystemTable() && y.isSystemTable()){
+            return (x.getVersion() == y.getVersion());
+        } else if (x.isSharedTable() && y.isSharedTable()) {
+            return (x.getName() == y.getName() & (x.getVersion()==y.getVersion()));
+        } else if (x.isLocalTable() && y.isLocalTable()){
+            SymbolTable[] x_table = x.getImportedTables();
+            SymbolTable[] y_table = x.getImportedTables();
+            if(x_table.length != y_table.length){
+                for(int i=0; i<x_table.length; i++){
+                    if(!isSameSymbolTable(x_table[i], y_table[i]))  return false;
+                }
+            }
+            return true;
+        } else if (x.isSubstitute() && y.isSubstitute()){
+            //TODO
+            return true;
```

Comment:
I think you can drop this branch. I don't believe it can be hit anyway, as substitute symbol tables are always shared symbol tables.

Reply:
deleted